org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/corrections/proposals/NewCUProposal.java

```diff
@@ -391,22 +390,32 @@ private String constructCUContent(ICompilationUnit cu, String typeContent, Strin
 	 * Called from createType to construct the source for this type
 	 */
 	private String constructTypeStub(ICompilationUnit parentCU, String name, int modifiers, String lineDelimiter) throws CoreException {
-		StringBuffer buf = new StringBuffer();
+		StringBuilder buf = new StringBuilder();
 
 		buf.append(Flags.toString(modifiers));
 		if (modifiers != 0) {
 			buf.append(' ');
 		}
+
+		String[] permittedNames = fCompilationUnit.findPrimaryType().getPermittedSubtypeNames();
+		boolean isPermitted = Arrays.asList(permittedNames).stream().anyMatch(p -> name.equals(p));
+		if (isPermitted) {
+			buf.append("final ");
+		}
+
 		String type = ""; //$NON-NLS-1$
 		String templateID = ""; //$NON-NLS-1$
+		String superType = ""; //$NON-NLS-1$
 		switch (fTypeKind) {
 			case K_CLASS:
 				type = "class "; //$NON-NLS-1$
 				templateID = CodeGeneration.CLASS_BODY_TEMPLATE_ID;
+				superType = fCompilationUnit.findPrimaryType().isInterface() ? "implements " : "extends ";
```

Comment:
we could avoid calling fCompilationUnit.findPrimaryType() several times in the same method, since a bunch of processing is happening behind the scene every time.